### PR TITLE
Add workID to payload, refactor packages, move NewUpkeepPayload from plugin to core

### PR DESCRIPTION
**Pull request by shileiwill**
This PR is in draft mode for discussion only.

In this PR:
1. I added a few util functions in upkeep_info.go: NewUpkeepPayload, PackTrigger, UnpackTrigger and UpkeepWorkID. 
2. The above functions are used to construct payload, workID from various places including evm and logprovider packages.

Some open questions:
1. provider.go is in logprovider package. If we import the functions from upkeep_info.go(in evm pkg), it will invoke circular dependencies. 
2. In transmit_event_provider.go, we need to unpack the trigger bytes to trigger object, need to confirm  if using the triggerID `UnpackTrigger(l.Id(), triggerID[:])` is sufficient. Also, to generate workID, need to manually convert trigger wrapper to trigger.

### CODE REVIEW

File: core/services/ocr2/plugins/ocr2keeper/evm21/encoder.go [hunk 1]
```diff
@@ -133,9 +133,12 @@ func (enc EVMAutomationEncoder21) Extract(raw []byte) ([]ocr2keepers.ReportedUpk
 			common.BytesToHash(triggerW.BlockHash[:]).Hex(),
 			logExt,
 		)
-		triggerID, _ := UpkeepTriggerID(upkeepId, report.Triggers[i])
+		id, _ := UpkeepTriggerID(upkeepId, report.Triggers[i])
+		// TODO workID should exclude the checkblock number
+		workId, _ := UpkeepTriggerID(upkeepId, report.Triggers[i])
```
Comment (shileiwill) [inline core/services/ocr2/plugins/ocr2keeper/evm21/encoder.go:138]: @EasterTheBunny Question: how to get rid of checkblock number here? Thanks.

Comment (EasterTheBunny) [reply core/services/ocr2/plugins/ocr2keeper/evm21/encoder.go:138]: You will probably need to create a new helper function for the WorkID.

File: core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/provider.go [hunk 2]
```diff
@@ -163,7 +163,14 @@ func (p *logEventProvider) GetLogs(context.Context) ([]ocr2keepers.UpkeepPayload
 			p.lggr.Warnw("failed to pack log data", "err", err, "log", log)
 			continue
 		}
-		payload := ocr2keepers.NewUpkeepPayload(l.id, logTriggerType, ocr2keepers.BlockKey(fmt.Sprintf("%d", log.BlockNumber)), trig, checkData)
+
+		// TODO: we can't import evm package to here because of circular dependency
+		payload, err := NewUpkeepPayload(l.id, logTriggerType, trig, checkData)
```
Comment (shileiwill) [inline core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/provider.go:168]: NewUpkeepPayload is in evm pkg, while this provider.go is in a different pkg. They have circular dependencies.

File: core/services/ocr2/plugins/ocr2keeper/evm21/transmit_event_provider.go [hunk 3]
```diff
@@ -211,12 +211,21 @@ func (c *TransmitEventProvider) convertToTransmitEvents(logs []transmitEventLog,
 	for _, l := range logs {
 		upkeepId := ocr2keepers.UpkeepIdentifier(l.Id().Bytes())
 		triggerID := l.TriggerID()
+		//TODO: is this the right way to get the trigger bytes?
+		//Unpack the trigger, triggerID is byte[32]
+		//triggerW, _ := UnpackTrigger(l.Id(), triggerID[:])
+
+		// TODO: need to manually convert trigger wrapper to trigger
```
Comment (shileiwill) [inline core/services/ocr2/plugins/ocr2keeper/evm21/transmit_event_provider.go:218]: As said in TODOs, first need to confirm how to construct Trigger object from a log. Second is to convert Trigger Wrapper to Trigger object which we can use to generate workID.

File: core/services/ocr2/plugins/ocr2keeper/evm21/encoder.go [hunk 4]
```diff
@@ -133,9 +133,11 @@ func (enc EVMAutomationEncoder21) Extract(raw []byte) ([]ocr2keepers.ReportedUpk
 			common.BytesToHash(triggerW.BlockHash[:]).Hex(),
 			logExt,
 		)
-		triggerID, _ := UpkeepTriggerID(upkeepId, report.Triggers[i])
+		id, _ := UpkeepTriggerID(upkeepId, report.Triggers[i])
+		workID, _ := UpkeepWorkID(upkeepId, trigger)
 		reportedUpkeeps[i] = ocr2keepers.ReportedUpkeep{
-			ID:          triggerID,
+			ID:          id,
+			WorkID:      workID,
```
Comment (shileiwill) [inline core/services/ocr2/plugins/ocr2keeper/evm21/encoder.go:140]: This is where we add WorkID to ReportedUpkeep.

File: go.mod [hunk 5]
```diff
@@ -71,7 +71,7 @@ require (
 	github.com/smartcontractkit/chainlink-solana v1.0.3-0.20230802143301-165000751a85
 	github.com/smartcontractkit/chainlink-starknet/relayer v0.0.1-beta-test.0.20230802150127-d2c95679d61a
 	github.com/smartcontractkit/libocr v0.0.0-20230724105022-c5e49b4195ac
-	github.com/smartcontractkit/ocr2keepers v0.7.2
+	github.com/smartcontractkit/ocr2keepers v1.0.0-alpha.0.0.20230801204339-7fd495548280
```
Comment (shileiwill) [inline go.mod:74]: will replace the alpha version before merge.

**Review by infiloop2 — CHANGES_REQUESTED** [review]
Since this is a cross repo change, we'll need to sync with other cross repo dev happening. Let's wait for https://github.com/smartcontractkit/chainlink/pull/10060 to be merged before merging this. (Requesting changes to prevent accidental merge)

**Comment by infiloop2** [comment]
Closing in favor of https://github.com/smartcontractkit/chainlink/pull/10095

**Pull request closed without merging**